core/src/main/java/ai/timefold/solver/core/impl/move/streams/maybeapi/generic/provider/ListChangeMoveProvider.java

```diff
@@ -0,0 +1,118 @@
+        };
+        this.validChangeFilter = (solutionView, value, targetPosition) -> {
+            var currentPosition = solutionView.getPositionOf(variableMetaModel, value);
+            if (currentPosition.equals(targetPosition)) { // No change needed.
+                return false;
+            }
+            if (currentPosition instanceof UnassignedElement) {
+                // Only assign the value if the target entity will accept it.
+                var targetPositionInList = targetPosition.ensureAssigned();
+                return solutionView.isValueInRange(variableMetaModel, targetPositionInList.entity(), value);
+            } else {
+                if (!(targetPosition instanceof PositionInList targetPositionInList)) { // Unassigning a value.
+                    return true;
+                }
+                var currentPositionInList = currentPosition.ensureAssigned();
+                if (currentPositionInList.entity() == targetPositionInList.entity()) {
+                    var valueCount = solutionView.countValues(variableMetaModel, currentPositionInList.entity());
+                    if (valueCount == 1) {
+                        return false; // The value is already in the list, and it is the only one.
+                    } else if (targetPositionInList.index() == valueCount) {
+                        // The value is already in the list, and we are trying to move it past the end of the list.
+                        return false;
+                    } else { // Same list, same position; ignore.
+                        return currentPositionInList.index() != targetPositionInList.index();
+                    }
+                } else { // We can move freely between entities, assuming the target entity accepts the value.
+                    return solutionView.isValueInRange(variableMetaModel, targetPositionInList.entity(), value);
+                }
+            }
+        };
+    }
+
+    @Override
+    public MoveProducer<Solution_> apply(MoveStreamFactory<Solution_> moveStreamFactory) {
+        // For each unassigned value, we need to create a move to assign it to some position of some list variable.
+        // For each assigned value that is not pinned, we need to create:
+        // - A move to unassign it.
+        // - A move to reassign it to another position if assigned.
+        // To assign or reassign a value, we need to create:
+        // - A move for every unpinned value in every entity's list variable to assign the value before that position.
+        // - A move for every entity to assign it to the last position in the list variable.
+        var unpinnedEntities =
+                moveStreamFactory.enumerate(variableMetaModel.entity().type(), variableMetaModel.allowsUnassignedValues());
+        var unpinnedValues = moveStreamFactory.enumerate(variableMetaModel.type(), true)
+                .filter((solutionView, value) -> value == null
```

Comment:
Does `value == null` mean unassigned?

Reply:
At this point in time, it just means that null value needs to survive until later.